// ews-javascript-api/js/ComplexProperties/FolderId.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var EwsUtilities_1 = require("../Core/EwsUtilities");
var ExtensionMethods_1 = require("../ExtensionMethods");
var WellKnownFolderName_1 = require("../Enumerations/WellKnownFolderName");
var XmlAttributeNames_1 = require("../Core/XmlAttributeNames");
var XmlElementNames_1 = require("../Core/XmlElementNames");
var ServiceId_1 = require("./ServiceId");
/**
 * Represents the Id of a folder.
 *
 * @class FolderId
 * @extends {ServiceId}
 */
var FolderId = /** @class */ (function (_super) {
    __extends(FolderId, _super);
    function FolderId(uniqueIdOrFolderName, mailbox) {
        var _this = this;
        arguments.length === 1 && typeof uniqueIdOrFolderName === 'string' ? _this = _super.call(this, uniqueIdOrFolderName) || this : _this = _super.call(this) || this;
        if (arguments.length > 0 && typeof uniqueIdOrFolderName === 'number') {
            _this.folderName = uniqueIdOrFolderName;
        }
        if (arguments.length > 1) {
            _this.mailbox = mailbox;
        }
        return _this;
    }
    Object.defineProperty(FolderId.prototype, "FolderName", {
        /**
         * Gets the name of the folder associated with the folder Id. Name and Id are mutually exclusive; if one is set, the other is null.
         *
         * @readonly
         * @type {WellKnownFolderName}
         */
        get: function () {
            return this.folderName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FolderId.prototype, "Mailbox", {
        /**
         * Gets the mailbox of the folder. Mailbox is only set when FolderName is set.
         *
         * @readonly
         * @type {Mailbox}
         */
        get: function () {
            return this.mailbox;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FolderId.prototype, "IsValid", {
        /**
         * True if this instance is valid, false otherthise.
         *
         * @value   *true* if this instance is valid; otherwise, *false*.
         */
        get: function () {
            if (ExtensionMethods_1.hasValue(this.FolderName)) {
                return (this.Mailbox == null) || this.Mailbox.IsValid;
            }
            else {
                return _super.prototype.IsValidProxy.call(this);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Determines whether the specified *FolderId* is equal to the current *FolderId*.
     * We do not consider the ChangeKey for FolderId.Equals.
     *
     * @param   {any}       obj   The  to compare with the current .
     * @return  {boolean}   true if the specified  is equal to the current ; otherwise, false.
     */
    FolderId.prototype.Equals = function (obj) {
        if (this === obj) {
            return true;
        }
        else {
            var other = obj;
            if (!(other instanceof FolderId)) {
                return false;
            }
            else if (this.FolderName) {
                if (other.FolderName && this.FolderName === other.FolderName) {
                    if (this.Mailbox != null) {
                        return this.Mailbox.Equals(other.Mailbox);
                    }
                    else if (other.Mailbox == null) {
                        return true;
                    }
                }
            }
            else if (_super.prototype.Equals.call(this, other)) {
                return true;
            }
            return false;
        }
    };
    //GetHashCode(): number { throw new Error("FolderId.ts - GetHashCode : Not implemented."); }
    /**
     * @internal Gets the name of the XML element.
     *
     * @return  {string}      XML element name.
     */
    FolderId.prototype.GetXmlElementName = function () {
        return typeof this.folderName !== 'undefined' && this.FolderName >= 0 ? XmlElementNames_1.XmlElementNames.DistinguishedFolderId : XmlElementNames_1.XmlElementNames.FolderId;
    };
    /**
     * Returns a *String* that represents the current *FolderId*.
     *
     * @return  {string}      A *String* that represents the current *FolderId*.
     */
    FolderId.prototype.ToString = function () {
        if (this.IsValid) {
            if (ExtensionMethods_1.hasValue(this.FolderName)) {
                if ((this.mailbox != null) && this.mailbox.IsValid) {
                    return ExtensionMethods_1.StringHelper.Format("{0} ({1})", WellKnownFolderName_1.WellKnownFolderName[this.folderName], this.Mailbox.ToString());
                }
                else {
                    return WellKnownFolderName_1.WellKnownFolderName[this.FolderName];
                }
            }
            else {
                return _super.prototype.ToString.call(this);
            }
        }
        else {
            return "";
        }
    };
    /**
     * @internal Validates FolderId against a specified request version.
     *
     * @param   {ExchangeVersion}   version   The version.
     */
    FolderId.prototype.Validate = function (version) {
        if (ExtensionMethods_1.hasValue(version)) {
            // The FolderName property is a WellKnownFolderName, an enumeration type. If the property
            // is set, make sure that the value is valid for the request version.
            if (ExtensionMethods_1.hasValue(this.FolderName)) {
                EwsUtilities_1.EwsUtilities.ValidateEnumVersionValue(WellKnownFolderName_1.WellKnownFolderName, this.FolderName, version, "WellKnownFolderName");
            }
        }
        else {
            _super.prototype.Validate.call(this);
        }
    };
    /**
     * @internal Writes attributes to XML.
     *
     * @param   {EwsServiceXmlWriter}   writer   The writer.
     */
    FolderId.prototype.WriteAttributesToXml = function (writer) {
        if (typeof this.folderName !== 'undefined' && this.FolderName >= 0) {
            writer.WriteAttributeValue(XmlAttributeNames_1.XmlAttributeNames.Id, WellKnownFolderName_1.WellKnownFolderName[this.FolderName].toLowerCase());
            if (this.Mailbox != null) {
                this.Mailbox.WriteToXml(writer, XmlElementNames_1.XmlElementNames.Mailbox);
            }
        }
        else {
            _super.prototype.WriteAttributesToXml.call(this, writer);
        }
    };
    return FolderId;
}(ServiceId_1.ServiceId));
exports.FolderId = FolderId;
