// ews-javascript-api/js/Autodiscover/Responses/GetUserSettingsResponse.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var AlternateMailboxCollection_1 = require("../AlternateMailboxCollection");
var AltDictionary_1 = require("../../AltDictionary");
var DocumentSharingLocationCollection_1 = require("../DocumentSharingLocationCollection");
var EwsLogging_1 = require("../../Core/EwsLogging");
var ProtocolConnectionCollection_1 = require("../ProtocolConnectionCollection");
var ExtensionMethods_1 = require("../../ExtensionMethods");
var UserSettingError_1 = require("../UserSettingError");
var UserSettingName_1 = require("../../Enumerations/UserSettingName");
var WebClientUrlCollection_1 = require("../WebClientUrlCollection");
var XmlElementNames_1 = require("../../Core/XmlElementNames");
var AutodiscoverResponse_1 = require("./AutodiscoverResponse");
/**
 * Represents the response to a GetUsersSettings call for an individual user.
 * @sealed
 */
var GetUserSettingsResponse = /** @class */ (function (_super) {
    __extends(GetUserSettingsResponse, _super);
    /**
     * Initializes a new instance of the **GetUserSettingsResponse** class.
     */
    function GetUserSettingsResponse() {
        var _this = _super.call(this) || this;
        _this.smtpAddress = null;
        _this.redirectTarget = null;
        _this.settings = null;
        _this.userSettingErrors = null;
        _this.smtpAddress = ExtensionMethods_1.StringHelper.Empty;
        _this.settings = new AltDictionary_1.Dictionary(function (value) { return value.toString(); });
        _this.userSettingErrors = [];
        return _this;
    }
    Object.defineProperty(GetUserSettingsResponse.prototype, "SmtpAddress", {
        /**
         * Gets the SMTP address this response applies to.
         */
        get: function () {
            return this.smtpAddress;
        },
        /** @internal */
        set: function (value) {
            this.smtpAddress = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GetUserSettingsResponse.prototype, "RedirectTarget", {
        /**
         * Gets the redirectionTarget (URL or email address)
         */
        get: function () {
            return this.redirectTarget;
        },
        /** @internal */
        set: function (value) {
            this.redirectTarget = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GetUserSettingsResponse.prototype, "Settings", {
        /**
         * Gets the requested settings for the user.
         */
        get: function () {
            return this.settings;
        },
        /** @internal */
        set: function (value) {
            this.settings = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GetUserSettingsResponse.prototype, "UserSettingErrors", {
        /**
         * Gets error information for settings that could not be returned.
         */
        get: function () {
            return this.userSettingErrors;
        },
        /** @internal */
        set: function (value) {
            this.userSettingErrors = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @internal Load from XML jsObject.
     *
     * @param   {any} jsObject  Json Object converted from XML.
     */
    GetUserSettingsResponse.prototype.LoadFromXmlJsObject = function (jsObject) {
        _super.prototype.LoadFromXmlJsObject.call(this, jsObject);
        var settingscol = jsObject[XmlElementNames_1.XmlElementNames.UserSettings];
        this.LoadUserSettingsFromXmlJsObject(settingscol);
        this.redirectTarget = jsObject[XmlElementNames_1.XmlElementNames.RedirectTarget] || null;
        this.LoadUserSettingErrorsFromXmlJsObject(jsObject[XmlElementNames_1.XmlElementNames.UserSettingErrors]);
    };
    /**
     * Loads the user setting errors.
     *
     * @param   {any} jsObject  Json Object converted from XML.
     */
    GetUserSettingsResponse.prototype.LoadUserSettingErrorsFromXmlJsObject = function (jsObject) {
        var errors = undefined;
        if (ExtensionMethods_1.isNullOrUndefined(jsObject) || ExtensionMethods_1.isNullOrUndefined(jsObject[XmlElementNames_1.XmlElementNames.UserSettingError]))
            return;
        if (Array.isArray(jsObject[XmlElementNames_1.XmlElementNames.UserSettingError]))
            errors = jsObject[XmlElementNames_1.XmlElementNames.UserSettingError];
        else
            errors = [jsObject[XmlElementNames_1.XmlElementNames.UserSettingError]];
        for (var i = 0; i < errors.length; i++) {
            var error = new UserSettingError_1.UserSettingError();
            error.LoadFromXmlJsObject(errors[i]);
            this.UserSettingErrors.push(error);
        }
    };
    /**
     * @internal Loads setting from XML jsObject.
     *
     * @param   {any} jsObject  Json Object converted from XML.
     */
    GetUserSettingsResponse.prototype.LoadUserSettingsFromXmlJsObject = function (jsObject) {
        var settings = undefined;
        if (!jsObject || typeof (jsObject[XmlElementNames_1.XmlElementNames.UserSetting]) === 'undefined')
            return;
        if (Object.prototype.toString.call(jsObject[XmlElementNames_1.XmlElementNames.UserSetting]) === "[object Array]")
            settings = jsObject[XmlElementNames_1.XmlElementNames.UserSetting];
        else
            settings = [jsObject[XmlElementNames_1.XmlElementNames.UserSetting]];
        for (var i = 0; i < settings.length; i++) {
            var setting = settings[i];
            var settingClass = setting["type"];
            switch (settingClass) {
                case XmlElementNames_1.XmlElementNames.StringSetting:
                case XmlElementNames_1.XmlElementNames.WebClientUrlCollectionSetting:
                case XmlElementNames_1.XmlElementNames.AlternateMailboxCollectionSetting:
                case XmlElementNames_1.XmlElementNames.ProtocolConnectionCollectionSetting:
                case XmlElementNames_1.XmlElementNames.DocumentSharingLocationCollectionSetting:
                    this.ReadSettingFromXmlJsObject(setting);
                    break;
                default:
                    EwsLogging_1.EwsLogging.Assert(false, "GetUserSettingsResponse.LoadUserSettingsFromXml", ExtensionMethods_1.StringHelper.Format("Invalid setting class '{0}' returned", settingClass));
                    break;
            }
        }
    };
    /**
     * Reads user setting from XML jsObject.
     *
     * @param   {any} jsObject  Json Object converted from XML.
     */
    GetUserSettingsResponse.prototype.ReadSettingFromXmlJsObject = function (jsObject) {
        var name = jsObject[XmlElementNames_1.XmlElementNames.Name];
        var value = jsObject[XmlElementNames_1.XmlElementNames.Value];
        switch (jsObject["type"]) {
            case XmlElementNames_1.XmlElementNames.WebClientUrlCollectionSetting: //.WebClientUrls:
                value = WebClientUrlCollection_1.WebClientUrlCollection.LoadFromXmlJsObject(jsObject[XmlElementNames_1.XmlElementNames.WebClientUrls]);
                break;
            case XmlElementNames_1.XmlElementNames.ProtocolConnectionCollectionSetting: //ProtocolConnections:
                value = ProtocolConnectionCollection_1.ProtocolConnectionCollection.LoadFromXmlJsObject(jsObject[XmlElementNames_1.XmlElementNames.ProtocolConnections]);
                break;
            case XmlElementNames_1.XmlElementNames.AlternateMailboxCollectionSetting: //AlternateMailboxes:
                value = AlternateMailboxCollection_1.AlternateMailboxCollection.LoadFromXmlJsObject(jsObject[XmlElementNames_1.XmlElementNames.AlternateMailboxes]);
                break;
            case XmlElementNames_1.XmlElementNames.DocumentSharingLocationCollectionSetting: //DocumentSharingLocations:
                //debugger;
                EwsLogging_1.EwsLogging.Log("------------DocumentSharingLocationCollection needs test and fix ----------------", true);
                EwsLogging_1.EwsLogging.Log(jsObject, true, true);
                value = DocumentSharingLocationCollection_1.DocumentSharingLocationCollection.LoadFromXmlJsObject(jsObject);
                break;
        }
        // EWS Managed API is broken with AutoDSvc endpoint in RedirectUrl scenario
        var userSettingName = UserSettingName_1.UserSettingName[name]; // EwsUtilities.Parse<UserSettingName>(name);
        if (userSettingName !== undefined)
            this.Settings.Add(userSettingName, value);
        else
            EwsLogging_1.EwsLogging.Assert(false, "GetUserSettingsResponse.ReadSettingFromXml", "Unexpected or empty name element in user setting");
    };
    /**
     * Tries the get the user setting value.
     *
     * @param   {UserSettingName}   setting   The setting.
     * @return  {T}  The setting value.
     */
    GetUserSettingsResponse.prototype.GetSettingValue = function (setting) {
        //public bool TryGetSettingValue<T>(UserSettingName setting, out T value)
        return this.Settings.get(setting);
    };
    return GetUserSettingsResponse;
}(AutodiscoverResponse_1.AutodiscoverResponse));
exports.GetUserSettingsResponse = GetUserSettingsResponse;
