// ngx-pipes/esm5/src/pipes/string/a-or-an.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { isVowel } from '../helpers/helpers';
var AorAnPipe = /** @class */ (function () {
    function AorAnPipe() {
        this.irregularMap = {
            herb: 'an',
            honor: 'an',
            honorable: 'an',
            hour: 'an',
            mba: 'an',
            msc: 'an',
            'm.sc.': 'an',
            unicorn: 'a',
        };
    }
    /**
     * @param {?} stringEntity
     * @return {?}
     */
    AorAnPipe.prototype.transform = /**
     * @param {?} stringEntity
     * @return {?}
     */
    function (stringEntity) {
        if (!stringEntity || stringEntity === '') {
            return '';
        }
        else {
            /** @type {?} */
            var firstWord = stringEntity.trim().split(' ')[0];
            if (this.irregularMap[firstWord.toLocaleLowerCase()]) {
                return this.irregularMap[firstWord.toLocaleLowerCase()] + " " + stringEntity;
            }
            else {
                return isVowel(stringEntity[0]) ? "an " + stringEntity : "a " + stringEntity;
            }
        }
    };
    AorAnPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'aOrAn',
                },] }
    ];
    return AorAnPipe;
}());
export { AorAnPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AorAnPipe.prototype.irregularMap;
}
