// ngx-pipes/esm2015/src/pipes/helpers/helpers.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} value
 * @return {?}
 */
export function isUndefined(value) {
    return typeof value === 'undefined';
}
/**
 * @param {?} value
 * @return {?}
 */
export function isFunction(value) {
    return typeof value === 'function';
}
/**
 * @param {?} value
 * @return {?}
 */
export function isNumber(value) {
    return typeof value === 'number';
}
/**
 * @param {?} value
 * @return {?}
 */
export function isString(value) {
    return typeof value === 'string';
}
/**
 * @param {?} value
 * @return {?}
 */
export function isBoolean(value) {
    return typeof value === 'boolean';
}
/**
 * @param {?} value
 * @return {?}
 */
export function isObject(value) {
    return value !== null && typeof value === 'object';
}
/**
 * @param {?} value
 * @return {?}
 */
export function isNumberFinite(value) {
    return isNumber(value) && isFinite(value);
}
/**
 * @param {?} letter
 * @return {?}
 */
export function isVowel(letter) {
    /** @type {?} */
    const vowels = ['a', 'e', 'i', 'o', 'u'];
    return vowels.indexOf(letter) !== -1;
}
/**
 * @param {?} text
 * @return {?}
 */
export function ucFirst(text) {
    const [part, ...split] = text.split(/\s/g);
    /** @type {?} */
    const ucd = part
        .toLowerCase()
        .split(/(?=['|-])/g)
        .map((/**
     * @param {?} word
     * @return {?}
     */
    (word) => word.indexOf('-') + word.indexOf("'") > -2
        ? word.slice(0, 2).toUpperCase() + word.slice(2)
        : word.slice(0, 1).toUpperCase() + word.slice(1)))
        .join('');
    return [ucd, ...split].join(' ');
}
/**
 * @param {?} num
 * @param {?} precision
 * @return {?}
 */
export function applyPrecision(num, precision) {
    if (precision <= 0) {
        return Math.round(num);
    }
    /** @type {?} */
    const tho = Math.pow(10, precision);
    return Math.round(num * tho) / tho;
}
/**
 * @param {?} obj
 * @param {?} map
 * @return {?}
 */
export function extractDeepPropertyByMapKey(obj, map) {
    /** @type {?} */
    const keys = map.split('.');
    /** @type {?} */
    const head = keys.shift();
    return keys.reduce((/**
     * @param {?} prop
     * @param {?} key
     * @return {?}
     */
    (prop, key) => {
        return !isUndefined(prop) && !isUndefined(prop[key]) ? prop[key] : undefined;
    }), obj[head || '']);
}
/**
 * @param {?} obj
 * @param {?} map
 * @return {?}
 */
export function extractDeepPropertyByParentMapKey(obj, map) {
    /** @type {?} */
    const keys = map.split('.');
    /** @type {?} */
    const tail = keys.pop();
    /** @type {?} */
    const props = extractDeepPropertyByMapKey(obj, keys.join('.'));
    return { props, tail };
}
/**
 * @param {?} obj
 * @param {?} other
 * @return {?}
 */
export function getKeysTwoObjects(obj, other) {
    return [...Object.keys(obj), ...Object.keys(other)].filter((/**
     * @param {?} key
     * @param {?} index
     * @param {?} array
     * @return {?}
     */
    (key, index, array) => array.indexOf(key) === index));
}
/**
 * @param {?} obj
 * @param {?} other
 * @return {?}
 */
export function isDeepEqual(obj, other) {
    if (!isObject(obj) || !isObject(other)) {
        return obj === other;
    }
    return getKeysTwoObjects(obj, other).every((/**
     * @param {?} key
     * @return {?}
     */
    (key) => {
        if (!isObject(obj[key]) && !isObject(other[key])) {
            return obj[key] === other[key];
        }
        if (!isObject(obj[key]) || !isObject(other[key])) {
            return false;
        }
        return isDeepEqual(obj[key], other[key]);
    }));
}
