// ews-javascript-api/js/Autodiscover/Requests/GetDomainSettingsRequest.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var AutodiscoverErrorCode_1 = require("../../Enumerations/AutodiscoverErrorCode");
var DomainSettingName_1 = require("../../Enumerations/DomainSettingName");
var EwsUtilities_1 = require("../../Core/EwsUtilities");
var GetDomainSettingsResponseCollection_1 = require("../Responses/GetDomainSettingsResponseCollection");
var ExtensionMethods_1 = require("../../ExtensionMethods");
var Promise_1 = require("../../Promise");
var ServiceValidationException_1 = require("../../Exceptions/ServiceValidationException");
var Strings_1 = require("../../Strings");
var XmlElementNames_1 = require("../../Core/XmlElementNames");
var XmlNamespace_1 = require("../../Enumerations/XmlNamespace");
var AutodiscoverRequest_1 = require("./AutodiscoverRequest");
/**
 * @internal Represents a GetDomainSettings request.
 */
var GetDomainSettingsRequest = /** @class */ (function (_super) {
    __extends(GetDomainSettingsRequest, _super);
    /**
     * @internal Initializes a new instance of the **GetDomainSettingsRequest** class.
     *
     * @param   {AutodiscoverService}   service   Autodiscover service associated with this request.
     * @param   {Uri}                   url       URL of Autodiscover service.
     */
    function GetDomainSettingsRequest(service, url) {
        return _super.call(this, service, url) || this;
    }
    Object.defineProperty(GetDomainSettingsRequest.prototype, "Domains", {
        /**
         * @internal Gets or sets the domains.
         */
        get: function () {
            return this.domains;
        },
        set: function (value) {
            this.domains = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GetDomainSettingsRequest.prototype, "Settings", {
        /**
         * @internal Gets or sets the settings.
         */
        get: function () {
            return this.settings;
        },
        set: function (value) {
            this.settings = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GetDomainSettingsRequest.prototype, "RequestedVersion", {
        /**
         * @internal Gets or sets the RequestedVersion.
         */
        get: function () {
            return this.requestedVersion;
        },
        set: function (value) {
            this.requestedVersion = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @internal Creates the service response.
     *
     * @return  {AutodiscoverResponse}      AutodiscoverResponse
     */
    GetDomainSettingsRequest.prototype.CreateServiceResponse = function () {
        return new GetDomainSettingsResponseCollection_1.GetDomainSettingsResponseCollection();
    };
    /**
     * @internal Executes this instance.
     *
     * @return  {Promise<GetDomainSettingsResponseCollection>}      [description]
     */
    GetDomainSettingsRequest.prototype.Execute = function () {
        return __awaiter(this, void 0, Promise_1.Promise, function () {
            var responses;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.InternalExecute()];
                    case 1:
                        responses = _a.sent();
                        if (responses.ErrorCode == AutodiscoverErrorCode_1.AutodiscoverErrorCode.NoError) {
                            this.PostProcessResponses(responses);
                        }
                        return [2 /*return*/, responses];
                }
            });
        });
    };
    /**
     * Gets the name of the request XML element.
     *
     * @return  {string}      [description]
     */
    GetDomainSettingsRequest.prototype.GetRequestXmlElementName = function () {
        return XmlElementNames_1.XmlElementNames.GetDomainSettingsRequestMessage;
    };
    /**
     * @internal Gets the name of the response XML element.
     *
     * @return  {string}      [description]
     */
    GetDomainSettingsRequest.prototype.GetResponseXmlElementName = function () {
        return XmlElementNames_1.XmlElementNames.GetDomainSettingsResponseMessage;
    };
    /**
     * @internal Gets the WS-Addressing action name.
     *
     * @return  {string}      [description]
     */
    GetDomainSettingsRequest.prototype.GetWsAddressingActionName = function () {
        return GetDomainSettingsRequest.GetDomainSettingsActionUri;
    };
    /**
     * Post-process responses to GetDomainSettings.
     *
     * @param   {GetDomainSettingsResponseCollection}   responses   The GetDomainSettings responses.
     */
    GetDomainSettingsRequest.prototype.PostProcessResponses = function (responses) {
        // Note:The response collection may not include all of the requested domains if the request has been throttled.
        for (var index = 0; index < responses.Count; index++) {
            responses.Responses[index].Domain = this.Domains[index];
        }
    };
    /**
     * @internal Validates the request.
     */
    GetDomainSettingsRequest.prototype.Validate = function () {
        _super.prototype.Validate.call(this);
        if (this.settings.length == 0) {
            throw new ServiceValidationException_1.ServiceValidationException(Strings_1.Strings.InvalidAutodiscoverSettingsCount);
        }
        if (this.domains.length == 0) {
            throw new ServiceValidationException_1.ServiceValidationException(Strings_1.Strings.InvalidAutodiscoverDomainsCount);
        }
        for (var _i = 0, _a = this.domains; _i < _a.length; _i++) {
            var domain = _a[_i];
            if (ExtensionMethods_1.StringHelper.IsNullOrEmpty(domain)) {
                throw new ServiceValidationException_1.ServiceValidationException(Strings_1.Strings.InvalidAutodiscoverDomain);
            }
        }
    };
    /**
     * @internal Writes attributes to request XML.
     *
     * @param   {EwsServiceXmlWriter}   writer   The writer.
     */
    GetDomainSettingsRequest.prototype.WriteAttributesToXml = function (writer) {
        writer.WriteAttributeValue("xmlns", EwsUtilities_1.EwsUtilities.AutodiscoverSoapNamespacePrefix, EwsUtilities_1.EwsUtilities.AutodiscoverSoapNamespace);
    };
    /**
     * @internal Writes elements to request XML.
     *
     * @param   {EwsServiceXmlWriter}   writer   The writer.
     */
    GetDomainSettingsRequest.prototype.WriteElementsToXml = function (writer) {
        writer.WriteStartElement(XmlNamespace_1.XmlNamespace.Autodiscover, XmlElementNames_1.XmlElementNames.Request);
        writer.WriteStartElement(XmlNamespace_1.XmlNamespace.Autodiscover, XmlElementNames_1.XmlElementNames.Domains);
        for (var _i = 0, _a = this.Domains; _i < _a.length; _i++) {
            var domain = _a[_i];
            if (!ExtensionMethods_1.StringHelper.IsNullOrEmpty(domain)) {
                writer.WriteElementValue(XmlNamespace_1.XmlNamespace.Autodiscover, XmlElementNames_1.XmlElementNames.Domain, domain);
            }
        }
        writer.WriteEndElement(); //Domains
        writer.WriteStartElement(XmlNamespace_1.XmlNamespace.Autodiscover, XmlElementNames_1.XmlElementNames.RequestedSettings);
        for (var _b = 0, _c = this.Settings; _b < _c.length; _b++) {
            var setting = _c[_b];
            writer.WriteElementValue(XmlNamespace_1.XmlNamespace.Autodiscover, XmlElementNames_1.XmlElementNames.Setting, DomainSettingName_1.DomainSettingName[setting]);
        }
        writer.WriteEndElement(); //RequestedSettings
        if (ExtensionMethods_1.hasValue(this.requestedVersion)) {
            writer.WriteElementValue(XmlNamespace_1.XmlNamespace.Autodiscover, XmlElementNames_1.XmlElementNames.RequestedVersion, this.requestedVersion);
        }
        writer.WriteEndElement(); //Request
    };
    /**
     * Action Uri of Autodiscover.GetDomainSettings method.
     */
    GetDomainSettingsRequest.GetDomainSettingsActionUri = EwsUtilities_1.EwsUtilities.AutodiscoverSoapNamespace + "/Autodiscover/GetDomainSettings";
    return GetDomainSettingsRequest;
}(AutodiscoverRequest_1.AutodiscoverRequest));
exports.GetDomainSettingsRequest = GetDomainSettingsRequest;
