// ngx-joyride/esm5/lib/services/templates.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
var TemplatesService = /** @class */ (function () {
    function TemplatesService() {
    }
    TemplatesService.prototype.setPrevButton = function (template) {
        this._prevButton = template;
    };
    TemplatesService.prototype.getPrevButton = function () {
        return this._prevButton;
    };
    TemplatesService.prototype.setNextButton = function (template) {
        this._nextButton = template;
    };
    TemplatesService.prototype.getNextButton = function () {
        return this._nextButton;
    };
    TemplatesService.prototype.setDoneButton = function (template) {
        this._doneButton = template;
    };
    TemplatesService.prototype.getDoneButton = function () {
        return this._doneButton;
    };
    TemplatesService.prototype.setCounter = function (template) {
        this._counter = template;
    };
    TemplatesService.prototype.getCounter = function () {
        return this._counter;
    };
    TemplatesService = __decorate([
        Injectable()
    ], TemplatesService);
    return TemplatesService;
}());
export { TemplatesService };
